### ISSUE TITLE
Pasting of rich content in textareas

### ISSUE THREAD

**Issue by GaurangTandon** [GaurangTandon/ProKeys#41]
Topic went offtopic and previous issue ([archive](http://pastebin.com/1Qsrj245)) was solved.
Scroll below.

**Comment by GaurangTandon** [comment]
**Solution:**

Writing a new content editable element's code is tough. So, use existing free software widgeditor - https://code.google.com/archive/p/widgeditor/downloads

**Comment by GaurangTandon** [comment]
New editor TinyMce now integrated:

![image](https://cloud.githubusercontent.com/assets/6308683/19018496/33c08850-8881-11e6-811f-8cd8ef2dd524.png)

_(Yes I said to use WidgEditor but then it's iframe can't be removed so I thought why not use a better editor)_

It has fullscreen editing, spell check support and pretty much everything you'll desire (but it has obviously doubled the size of prokeys now :/)

**Comment by GaurangTandon** [comment]
# **Issue**:

Tinymce adds a lot of special formatting features. And I also intend to provide a textarea view (which is basically the text-only view with no special formatting). Note that if user writes cool effects, like this:

![image](https://cloud.githubusercontent.com/assets/6308683/21286524/a14bf7c4-c47c-11e6-9d74-0d021e5c1c14.png)
_(has a table and center aligned elements)_

and then uses textarea editor, he'll get this:

    <p></p><table class="mce-item-table" width="353" style="height: 97px;" data-mce-style="height: 97px;"><tbody><tr style="height: 13px;" data-mce-style="height: 13px;"><td style="width: 62px; height: 13px;" data-mce-style="width: 62px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 65px; height: 13px;" data-mce-style="width: 65px; height: 13px;"><br></td></tr><tr style="height: 13px;" data-mce-style="height: 13px;"><td style="width: 62px; height: 13px;" data-mce-style="width: 62px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 65px; height: 13px;" data-mce-style="width: 65px; height: 13px;"><br></td></tr><tr style="height: 13px;" data-mce-style="height: 13px;"><td style="width: 62px; height: 13px;" data-mce-style="width: 62px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 65px; height: 13px;" data-mce-style="width: 65px; height: 13px;"><br></td></tr><tr style="height: 13px;" data-mce-style="height: 13px;"><td style="width: 62px; height: 13px;" data-mce-style="width: 62px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;">asdasdasd</td><td style="width: 65px; height: 13px;" data-mce-style="width: 65px; height: 13px;"><br></td></tr><tr style="height: 13px;" data-mce-style="height: 13px;"><td style="width: 62px; height: 13px;" data-mce-style="width: 62px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 64px; height: 13px;" data-mce-style="width: 64px; height: 13px;"><br></td><td style="width: 65px; height: 13px;" data-mce-style="width: 65px; height: 13px;"><br data-mce-bogus="1"></td></tr></tbody></table><p>asdasd</p><p style="text-align: center;" data-mce-style="text-align: center;">s</p><p style="text-align: center;" data-mce-style="text-align: center;">dasdasd</p><p><br data-mce-bogus="1"></p><p>dasdasd<span style="text-decoration: underline;" data-mce-style="text-decoration: underline;">asdasdasdasd</span></p><p><span style="text-decoration: underline;" data-mce-style="text-decoration: underline;"><br data-mce-bogus="1"></span></p><p><span style="text-decoration: underline;" data-mce-style="text-decoration: underline;"><br data-mce-bogus="1"></span></p><div data-mce-bogus="all" class="mce-resize-bar mce-resize-bar-row" unselectable="on" data-mce-resize="false" data-mce-style="cursor: row-resize; margin: 0; padding: 0; position: absolute; left: 8px; top: 31px; height: 4px; width: 353px; " style="cursor: row-resize; margin: 0; padding: 0; position: absolute; left: 8px; top: 31px; height: 4px; width: 353px; " data-row="0"></div><div data-mce-bogus="all" class="mce-resize-bar mce-resize-bar-row" unselectable="on" data-mce-resize="false" data-mce-style="cursor: row-resize; margin: 0; padding: 0; position: absolute; left: 8px; top: 50px; height: 4px; width: 353px; " style="cursor: row-resize; margin: 0; padding: 0; position: absolute; left: 8px; top: 50px; height: 4px; width: 353px; " data-row="1"></div><div data-mce-bogus="all" class="mce-resize-bar mce-resize-bar-row" unselectable="on" data-mce-resize="false" data-mce-style="cursor: row-resize; margin: 0; padding: 0; position: absolute; left: 8px; top: 69px; height: 4px; width: 353px; " style="cursor: row-resize; margin: 0; padding: 0; position: absolute; left: 8px; top: 69px; height: 4px; width: 353px; " data-row="2"></div><div data-mce-bogus="all" class="mce-resize-bar mce-resize-bar-row" unselectable="on" data-mce-resize="false" data-mce-style="cursor: row-resize; margin: 0; padding: 0; position: absolute; left: 8px; top: 88px; height: 4px; width: 353px; " style="cursor: row-resize; margin: 0; padding: 0; position: absolute; left: 8px; top: 88px; height: 4px; width: 353px; " data-row="3"></div><div data-mce-bogus="all" class="mce-resize-bar mce-resize-bar-row" unselectable="on" data-mce-resize="false" data-mce-style="cursor: row-resize; margin: 0; padding: 0; position: absolute; left: 8px; top: 105px; height: 4px; width: 353px; " style="cursor: row-resize; margin: 0; padding: 0; position: absolute; left: 8px; top: 105px; height: 4px; width: 353px; " data-row="4"></div><div data-mce-bogus="all" class="mce-resize-bar mce-resize-bar-col" unselectable="on" data-mce-resize="false" data-mce-style="cursor: col-resize; margin: 0; padding: 0; position: absolute; left: 77px; top: 11px; height: 99px; width: 4px; " style="cursor: col-resize; margin: 0; padding: 0; position: absolute; left: 77px; top: 11px; height: 99px; width: 4px; " data-col="0"></div><div data-mce-bogus="all" class="mce-resize-bar mce-resize-bar-col" unselectable="on" data-mce-resize="false" data-mce-style="cursor: col-resize; margin: 0; padding: 0; position: absolute; left: 147px; top: 11px; height: 99px; width: 4px; " style="cursor: col-resize; margin: 0; padding: 0; position: absolute; left: 147px; top: 11px; height: 99px; width: 4px; " data-col="1"></div><div data-mce-bogus="all" class="mce-resize-bar mce-resize-bar-col" unselectable="on" data-mce-resize="false" data-mce-style="cursor: col-resize; margin: 0; padding: 0; position: absolute; left: 217px; top: 11px; height: 99px; width: 4px; " style="cursor: col-resize; margin: 0; padding: 0; position: absolute; left: 217px; top: 11px; height: 99px; width: 4px; " data-col="2"></div><div data-mce-bogus="all" class="mce-resize-bar mce-resize-bar-col" unselectable="on" data-mce-resize="false" data-mce-style="cursor: col-resize; margin: 0; padding: 0; position: absolute; left: 287px; top: 11px; height: 99px; width: 4px; " style="cursor: col-resize; margin: 0; padding: 0; position: absolute; left: 287px; top: 11px; height: 99px; width: 4px; " data-col="3"></div><div data-mce-bogus="all" class="mce-resize-bar mce-resize-bar-col" unselectable="on" data-mce-resize="false" data-mce-style="cursor: col-resize; margin: 0; padding: 0; position: absolute; left: 356px; top: 11px; height: 99px; width: 4px; " style="cursor: col-resize; margin: 0; padding: 0; position: absolute; left: 356px; top: 11px; height: 99px; width: 4px; " data-col="4"></div>

I can surely remove the `p` elements, but the table if out of my control. You've any suggestions, @iWrote ?

**Comment by iWrote** [comment]
# Possible Solutions. 

## **Summary -**

IF autoFormat takes too long to implement
THEN remove tables, add them later
ELSE implement autoformat

ALSO dual-save in completely unnecessary.


## Not summary.
Says the same thing. So I added the summary in an afterthought.

## **One.**
Most people will just use the main editor. Those who choose to use the textarea **after adding all those effects** will probably expect a flood of markup anyway, so they shouldn't mind it **IF** -- it's **formatted to be readable.** 

In its current state, it's unusable. If auto-format will take a while to implement see solution 2. 

### **Two.**
TinyMCE itself is on the verge of overkill. 
 I think we can safely **remove tables for now** and add them in a future version. 
**Strip the editor of all non-essential/problematic features and incorporate them in future versions.**

Our 3.1.0 deadline is spinning in its grave.

Has anybody been requesting tables? 
(They could wait.)

### About dual-save.
Doesn't formatted text revert to plain text if it's pasted into textareas?
**Users can just save two snippets -** _kekMsg_rich_ **and** _kekMsg_plain_ **- if they're so desperate.**

P.S. Sry for the weird numbering. Lists in markdown don't like me for some reason.

**Comment by GaurangTandon** [comment]
Awesome, I'll remove the tables. We've another problem now, suppose the user saves this rich text:

> **bold text rockssss**
> *I am awesome*

and then uses it on a textarea (like gh comments). Should ProKeys paste:

**Format A**
````
<b>bold text rocksss</b>
<i>I am awesome</i>
````

or just:

**Format B**
````
bold text rockssss
I am awesome
````

Note that Format A actually makes sense on sites like github or stackoverflow because even though they've normal comments boxes, the `<b>`, `<i>`, etc. tags actually get formatted to produce meaningful result.

(<i>Example result brought to you by `i` tags</i>)

**Comment by iWrote** [comment]
Unless you can detect relevance. **We must ask the users.** 


________________________________________________________________________________________________________________

**One.**  New setting-

> [ ] Paste with tags by default.
 Perhaps with a list to add particular sites. Like the "[ ] Disable prokeys on these websites -"

Will clutter Settings page.

________________________________________________________________________________________________________________





**Two.**  Pressing **Alt+Space** to do a PasteSpecial. 
ProKeys could permanently move to having **2 hotkeys. Regular and notRegular**. The latter could help with other **context sensitive** ambiguities in the future. 
(paste-without-formatting, ignore delimiter, doNotMath, doNotComplete . . .)

This change might have some **growth** potential. 


________________________________________________________________________________________________________________



**Three.**  Typing snip-prime / snip-dash / snipSuffix could paste without formatting -

```
 mySnip   ==> <i>My Snippet</i>         
 mySnip`  ==>  My Snippet
```

Backtick doesn't create an edge case. Present on most keyboards. We could use keyword suffixes too --     detect suffix -> identify snip before it -> magic.

Keywords also have growth potential. Could serve more purposes.


________________________________________________________________________________________________________________
We could do a meaningful system combining Two and Three. This would mean expanding ProKeys in a new direction, in a possibly unnecessary one. There's feature-rich, and feature-overloaded.

**Comment by GaurangTandon** [comment]
You did good analysis. Kudos! Let us proceed systematically:

> Unless you can detect relevance. We must ask the users.
No, I _cannot_ detect this relevance i.e. where `<b>` tag will get formatted and where it won't.

> One. New setting-

**REJECTED** Too much hardwork of implementation for me, new clutter in settings page and cognitive burden of settings for users to keep up with. 

> Two. Pressing Alt+Space to do a PasteSpecial
Having to keep track of multiple hotkeys is not as much a worry for me as it might be for users. But only a few pro users will use multiple hotkeys anyway so not a problem. However, one hotkey can _only_ do one thing. That means in case we introduce another mod in future, we'll have to introduce _another hotkey._

> Three. Typing snip-dash
This seems to be perfect. We can _reserve_ symbols like backtick (SUGGEST MORE HERE) so that they can be further used in the growth potential.

But typing a hotkey to PasteSpecial is seamless as compared typing snip-dash and then pressing the regular hotkey.

We need to think over what other special cases might arise where we might need this special hotkey/keySymbol:

1. **Paste without formatting:** paste text stripped of all bold, italics, etc. tags.
2. <s>REJECTED DoNotMath-does not make sense; math isn't done inside snippets; it is a separate feature</s>
3. ignoreDelimiter - what does that even mean @iWrote ?
4. and more?

If the count of future expansions doesn't exceed the current count, we might introduce separate hotkey options.

**Comment by GaurangTandon** [comment]
Has been fixed in #147. As of now, bold, italics, pre etc. tags are being retained while alignment, color, etc. styles are removed on pasting in textarea, as one would expect for markdown enabled fields.

**Comment by GaurangTandon** [comment]
I think this decision goes along the lines of "You saved the snippet using an html editor, and you attempted to paste it in a textarea, so we retained the most plausible formatting we could so that you would view it as you wanted. If you didn't want any formatting in the first place, better use the textarea instead"
